File: backend/core/api_models/template_submissions.py
"""Template submission API models."""

from enum import Enum
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class TemplateSubmissionStatus(str, Enum):
    """Status of a template submission."""
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class TemplateSubmissionCreateRequest(BaseModel):
    """Request model for creating a template submission."""
    agent_id: str = Field(..., description="The ID of the agent to submit as a template")
    template_name: str = Field(..., min_length=1, max_length=255, description="Name for the template")
    template_description: Optional[str] = Field(None, description="Description of the template")
    category_id: Optional[str] = Field(None, description="Category ID for the template")
    use_cases: Optional[List[str]] = Field(None, description="Example use cases for the template")


class TemplateSubmissionResponse(BaseModel):
    """Response model for a template submission."""
    submission_id: str
    agent_id: str
    submitter_id: str
    template_name: str
    template_description: Optional[str] = None
    category_id: Optional[str] = None
    use_cases: Optional[List[str]] = None
    status: TemplateSubmissionStatus
    submitted_at: str
    reviewed_at: Optional[str] = None
    reviewed_by: Optional[str] = None
    rejection_reason: Optional[str] = None
    published_template_id: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: str
    updated_at: str

    # Joined fields (optional, populated when needed)
    agent_name: Optional[str] = None
    submitter_email: Optional[str] = None
    category_name: Optional[str] = None


class TemplateSubmissionsListResponse(BaseModel):
    """Response model for a list of template submissions."""
    submissions: List[TemplateSubmissionResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class ApproveSubmissionRequest(BaseModel):
    """Request model for approving a template submission."""
    admin_notes: Optional[str] = Field(None, description="Internal notes from the admin")


class RejectSubmissionRequest(BaseModel):
    """Request model for rejecting a template submission."""
    rejection_reason: str = Field(..., min_length=1, description="Reason for rejection (sent to user)")
    admin_notes: Optional[str] = Field(None, description="Internal notes from the admin")


class TemplateSubmissionStatsResponse(BaseModel):
    """Response model for template submission statistics."""
    total_submissions: int
    pending_count: int
    approved_count: int
    rejected_count: int
    submissions_this_week: int
    avg_review_time_hours: Optional[float] = None

File: apps/frontend/src/lib/api/template-submissions.ts
/**
 * API client for template submissions.
 */

import { backendApi } from '../api-client';

// Types

export type TemplateSubmissionStatus = 'pending' | 'approved' | 'rejected';

export interface TemplateSubmissionCreateRequest {
  agent_id: string;
  template_name: string;
  template_description?: string;
  category_id?: string;
  use_cases?: string[];
}

export interface TemplateSubmission {
  submission_id: string;
  agent_id: string;
  submitter_id: string;
  template_name: string;
  template_description?: string;
  category_id?: string;
  use_cases?: string[];
  status: TemplateSubmissionStatus;
  submitted_at: string;
  reviewed_at?: string;
  reviewed_by?: string;
  rejection_reason?: string;
  published_template_id?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  // Joined fields
  agent_name?: string;
  submitter_email?: string;
  category_name?: string;
}

export interface TemplateSubmissionsListResponse {
  submissions: TemplateSubmission[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface TemplateSubmissionStatsResponse {
  total_submissions: number;
  pending_count: number;
  approved_count: number;
  rejected_count: number;
  submissions_this_week: number;
  avg_review_time_hours?: number;
}

export interface ApproveSubmissionRequest {
  admin_notes?: string;
}

export interface RejectSubmissionRequest {
  rejection_reason: string;
  admin_notes?: string;
}

// Helper to build query strings
function buildQueryString(params: Record<string, string | number | boolean | undefined>): string {
  const searchParams = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      searchParams.append(key, String(value));
    }
  });
  const queryString = searchParams.toString();
  return queryString ? `?${queryString}` : '';
}

// User API functions

export async function createTemplateSubmission(
  request: TemplateSubmissionCreateRequest
): Promise<TemplateSubmission> {
  const response = await backendApi.post<TemplateSubmission>('/template-submissions', request);
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to create template submission');
  }
  return response.data;
}

export async function listMySubmissions(params?: {
  status?: TemplateSubmissionStatus;
  page?: number;
  page_size?: number;
}): Promise<TemplateSubmissionsListResponse> {
  const queryString = buildQueryString(params ?? {});
  const response = await backendApi.get<TemplateSubmissionsListResponse>(
    `/template-submissions${queryString}`
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to list submissions');
  }
  return response.data;
}

export async function getSubmission(submissionId: string): Promise<TemplateSubmission> {
  const response = await backendApi.get<TemplateSubmission>(
    `/template-submissions/${submissionId}`
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to get submission');
  }
  return response.data;
}

export async function cancelSubmission(
  submissionId: string
): Promise<{ message: string }> {
  const response = await backendApi.delete<{ message: string }>(
    `/template-submissions/${submissionId}`
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to cancel submission');
  }
  return response.data;
}

// Admin API functions

export async function listAllSubmissions(params?: {
  status?: TemplateSubmissionStatus;
  page?: number;
  page_size?: number;
}): Promise<TemplateSubmissionsListResponse> {
  const queryString = buildQueryString(params ?? {});
  const response = await backendApi.get<TemplateSubmissionsListResponse>(
    `/admin/template-submissions${queryString}`
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to list all submissions');
  }
  return response.data;
}

export async function getSubmissionStats(): Promise<TemplateSubmissionStatsResponse> {
  const response = await backendApi.get<TemplateSubmissionStatsResponse>(
    '/admin/template-submissions/stats'
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to get submission stats');
  }
  return response.data;
}

export async function approveSubmission(
  submissionId: string,
  request?: ApproveSubmissionRequest
): Promise<TemplateSubmission> {
  const response = await backendApi.post<TemplateSubmission>(
    `/admin/template-submissions/${submissionId}/approve`,
    request ?? {}
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to approve submission');
  }
  return response.data;
}

export async function rejectSubmission(
  submissionId: string,
  request: RejectSubmissionRequest
): Promise<TemplateSubmission> {
  const response = await backendApi.post<TemplateSubmission>(
    `/admin/template-submissions/${submissionId}/reject`,
    request
  );
  if (!response.success || !response.data) {
    throw response.error || new Error('Failed to reject submission');
  }
  return response.data;
}
